src/rust/engine/fs/src/store.rs

```diff
@@ -603,33 +612,33 @@ mod remote {
           )).to_boxed() as BoxFuture<_, _>
         }
         Ok((sender, receiver)) => {
-          let mut offset = 0 as usize;
-          let chunks: Vec<_> = bytes
-            .chunks(self.chunk_size_bytes)
-            .into_iter()
-            .map(|chunk| {
-              let mut req = bazel_protos::bytestream::WriteRequest::new();
-              req.set_resource_name(resource_name.clone());
-              req.set_write_offset(offset as i64);
-              offset += chunk.len();
-              req.set_finish_write(offset == len);
-              req.set_data(chunk.to_vec());
-              (req, grpcio::WriteFlags::default())
-            })
-            .collect();
+          let chunk_size_bytes = self.chunk_size_bytes;
+          let stream =
+            futures::stream::unfold::<_, _, futures::future::FutureResult<_, grpcio::Error>, _>(
+              0 as usize,
+              move |offset| if offset >= bytes.len() {
+                None
+              } else {
+                let mut req = bazel_protos::bytestream::WriteRequest::new();
+                req.set_resource_name(resource_name.clone());
+                req.set_write_offset(offset as i64);
+                let next_offset = min(offset + chunk_size_bytes, bytes.len());
+                req.set_finish_write(next_offset == bytes.len());
+                req.set_data(bytes.slice(offset, next_offset).to_vec());
```

Comment:
It's possible to make this a bit faster by creating a `BytesMut` instance rather than a `Bytes` instance, and using the [split_to](https://carllerche.github.io/bytes/bytes/struct.BytesMut.html#method.split_to) method to give away views of the prefix without copying.

Reply:
That would require us to change how the protobuf generation is done. I'm totally on board with that, but I'll do it as a separate PR :)